nemo/collections/tts/modules/radtts.py

```diff
@@ -777,33 +805,71 @@ def _prepare_for_export(self, **kwargs):
         self.remove_norms()
         super()._prepare_for_export(**kwargs)
 
-    def input_example(self, max_batch=1, max_dim=256):
-        """
-        Generates input examples for tracing etc.
-        Returns:
-            A tuple of input examples.
-        """
+    def input_example(self, max_batch=1, max_dim=400):
         par = next(self.parameters())
         sz = (max_batch, max_dim)
-        inp = torch.randint(16, 32, sz, device=par.device, dtype=torch.int64)
-        lens = torch.randint(max_dim // 8, max_dim // 4, (max_batch,), device=par.device, dtype=torch.int)
-        zeros_src = torch.zeros_like(inp)
-        inp.scatter_(1, lens.long().unsqueeze(-1) - 1, zeros_src)  # set the inp[i, lens[i]-1] elt to zero
+        # sz = (max_batch * max_dim,)
+        inp = torch.randint(0, 94, sz, device=par.device, dtype=torch.int64)
         speaker = torch.randint(0, 1, (max_batch,), device=par.device, dtype=torch.int64)
+        pitch = torch.randn(sz, device=par.device, dtype=torch.float32) * 0.5
+        pace = torch.clamp(torch.randn(sz, device=par.device, dtype=torch.float32) * 0.1 + 1, min=0.01)
+        volume = torch.clamp(torch.randn(sz, device=par.device, dtype=torch.float32) * 0.1 + 1, min=0.01)
+        # batch_lengths = torch.zeros((max_batch + 1), device=par.device, dtype=torch.int32)
+        # left_over_size = sz[0]
+        # batch_lengths[0] = 0
+        # for i in range(1, max_batch):
+        #     length = torch.randint(1, left_over_size - (max_batch - i), (1,), device=par.device)
+        #     batch_lengths[i] = length + batch_lengths[i - 1]
+        #     left_over_size -= length.detach().cpu().numpy()[0]
+        # batch_lengths[-1] = left_over_size + batch_lengths[-2]
+        # sum = 0
+        # index = 1
+        # while index < len(batch_lengths):
+        #     sum += batch_lengths[index] - batch_lengths[index - 1]
+        #     index += 1
+        # assert sum == sz[0], f"sum: {sum}, sz: {sz[0]}, lengths:{batch_lengths}"
+
+        # TODO: Shouldn't hardcode but self.tokenizer isn't initlized yet so unsure how
+        # to get the pad_id
+        pad_id = 94
+        inp[inp == pad_id] = pad_id - 1 if pad_id > 0 else pad_id + 1
```

Comment:
@XuesongYang @redoctopus Any suggestions on this?

Reply:
When is `input_example()` called? Is it possible to create the tokenizer first?

One way around it might be to have the `pad_id` be manually calculated beforehand using number of total valid tokens (with a note that explains that it's how the tokenizer assigns the `pad_id`). I'd be somewhat worried about it causing a subtle bug in the future, though, if something changes and they no longer match.

Another possibility--maybe the user can specify it in the config? They'll have to figure out their vocab length first though (annoying user experience).

The most ideal thing is just to have the tokenizer in existence first, if possible...

Reply:
I would suggest preserving `<pad>` ID (along with `<unk>` and `<blank>`) strictly in either the beginning of the file `https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/common/tokenizers/text_to_speech/tts_tokenizers.py`, or `https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/common/tokenizers/text_to_speech/ipa_lexicon.py`, as a shared resource to all tokenizers. 

for general tokenization purposes, token IDs for `<pad>`, `<unk>`, and `<blank>` should be preserved ahead to avoid any confusion and potential bugs. For example, predefine some mapping `{"<pad>": 0, `<unk>`: 1, `<blank>`: 2}`. Dynamically assigning token IDs for those may bring maintenance overheads...

From my current experience inside NeMo, any mel-generator model (not module) should call `self._setup_tokenizer()` when instantiating itself, i.e. https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/tts/models/fastpitch.py#L93-L98. Instantiating g2p happens inside `self._setup_tokenizer()` as shown in https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/tts/models/fastpitch.py#L213.

Reply:
I agree with @XuesongYang  - tokenizer has to be initialized at the time of input_example call. And we should use ```<pad>``` token in config.

Reply:
You're right, it is instantiated in the model. Not sure why this functionality was defined in the module, but moving to from module to model solves this issue